Improve scaling, fix bugs and add logging in economicsAI

jaj22 opened the pull request:
## What type of PR is this.
1. [X] Bug
2. [X] Change
3. [ ] Enhancement

### What have you changed and why?
- Fixed the increment function so that it caps correctly for array inputs.
- Used linear player scaling that's basically identical to #2079 but less opaque. Fixed for HCs.
- Increased tier scaling because we don't use many fancy vehicles at low tiers.
- Added a bit of constant to the caps so that carrier-only attacks aren't full of junk.
- Balance scale the caps as well as the increment. Otherwise enemies can pile up vehicles with some playstyles.
- Tweaked the important unit type coefficients based on some community server data.
- Added the logging from #2079, reduced code.

It's normalized such that 9 players at 7 wartier gives one vehicle per hour at typeCoeff 1, or two vehicles per hour for 26 players. Compared to the previous code, the rates are similar for ~26 players but substantially lower for low player counts. This fits my other data point where halving the rate was relatively fair for 3-4 players. Note that in the original version, occupants and invaders destroyed each other's vehicles far more frequently due to lack of attack simulation.

Didn't bother thinking about the coeffs for static AA/AT & boats. I doubt they matter much.

### Please specify which Issue this PR Resolves.
closes #1162
supersedes some of #2079

### Please verify the following and ensure all checks are completed.
1. [X] Have you loaded the mission in LAN host?
2. [ ] Have you loaded the mission on a dedicated server?

### Is further testing or are further changes required?
1. [X] No
2. [ ] Yes (Please provide further detail below.)

It'll do until I scrap it fairly shortly.

### How can the changes be tested?
Run `call A3A_fnc_economicsAI` a few dozen times as server. Check what it throws at the logs.

CalebSerafin commented on A3-Antistasi/functions/Base/fn_economicsAI.sqf:
```diff
@@ -1,67 +1,84 @@
-[0.2, "", vehNATOMRLS, _airbases + _outposts * 0.2, _accelerator] spawn _fnc_economics;
+private _airCap = _balanceScale * (4 + _airbases*2);
+private _groundCap = _balanceScale * (4 + _airbases + _outposts*0.5);
+private _increase = _balanceScale / 6;      // Because it's called six times per hour. Could pass in a time instead...
```
Until the actual timing is imported, a note should be added to keep these values in sync. By the sleep counter in [fn_resourcecheck Ln 120](https://github.com/official-antistasi-community/A3-Antistasi/blob/dc7e93be3718f7dbd5edb768b388819687991812/A3-Antistasi/functions/init/fn_resourcecheck.sqf#L120).

Or instead, this function could be made frequency agnostic by keeping track of the last time it was called. Then it will function without needing to know how many times it should be called. This will be better as it would separate the concern of economicsAI from resourceCheck or any new function that calls it.

jaj22 replied:
I don't think the latter option is workable because resourceCheck intentionally pauses the economics when there's no commander on the server. Hence it's easier for resourceCheck to manage the call rate.

I don't see economicsAI outliving the 10-minute tick unless someone is planning to rewrite resourceCheck very shortly. Otherwise I'm more inclined to add a parameter than document it further.

CalebSerafin commented on A3-Antistasi/functions/Base/fn_economicsAI.sqf:
```diff
@@ -1,67 +1,84 @@
+[[vehNATOTank], 0.6, _groundCap, _increase] call _fnc_economics;
+[[vehNATOAA], 0.6, _groundCap, _increase] call _fnc_economics;
+[[vehNATOMRLS], 0.3, _groundCap, _increase] call _fnc_economics;           // not used atm?
+[[vehNATOBoat], 1.0, _balanceScale * (1 + _seaports), _increase] call _fnc_economics;
```
Seaports are a land base. So as the AI faction gets pushed off the island, they will end up with fewer boats. However, in that state, it might make sense to give the AI more boats to show the player the change in unit composition as the AI lose land bases. So seaports should rather affect inland forces. Players don't capture seaports because they perceive boats as a threat, instead, it's captured similar to other land bases.

BUT, I do not know whether the AI can spawn boats when they do not have a seaport, so there may be a technical limitation then.

jaj22 replied:
I think there are four ways that gunboats (vehNATOboat, vehCSATboat) can be spawned:
1. As a bonus to the first wave of wavedCA. Relies on seaspawn markers which are missing on some maps.
2. At seaports as static-ish defence.
3. Sea patrols. Not absolutely sure about these because I haven't seen the code. Might be wandering seaport spawns.
4. As part of the salvage mission. This one doesn't check the availability.

My experience is that the vast majority of the boats you see in practice are placed at seaports, so primarily scaling them by seaport is appropriate. I think that's about the limit to how much thought I'm willing to put into it at the moment.

CalebSerafin commented on A3-Antistasi/functions/Base/fn_economicsAI.sqf:
```diff
@@ -1,67 +1,84 @@
-[0.2, "random", vehCSATTransportHelis - [vehCSATPatrolHeli], _airbases * 4, _accelerator] spawn _fnc_economics;
-[0.2, "random", vehCSATAttackHelis, _airbases * 4, _accelerator] spawn _fnc_economics;
-[0.2, "", vehCSATMRLS, _airbases + _outposts * 0.2, _accelerator] spawn _fnc_economics;
+_airCap = 1.2 * _balanceScale * (4 + _airbases*2);
```
To show the intent: it would be better to explicitly take the Occupant's value and then multiply it by 1.2. Because future code changes could make the formulas become out of sync.

jaj22 replied:
Can't do that because the airbase/outpost count is different. Vars should really be `_occAirbases` etc for transparency, but I copied the old code for that.

What I would be inclined to do is pull the 1.2 and the implicit 1.0 out into global or faction variables so that users can modify them on the fly, but I don't really want to add global vars for a temporary concept. Could pull them to defines at least.

CalebSerafin commented on A3-Antistasi/functions/Base/fn_economicsAI.sqf:
```diff
@@ -1,67 +1,84 @@
-        if (_currentItems < _maxItems) then {
-            timer setVariable [selectRandom _typeX, _currentItems + _coefficient * _accelerator, true];
-        };
+    if (_totalItems < ceil (_typeCoeff * _baseCap)) then {
```
The ceiling here is unnecessary, as `_totalItems` will always be an integer, or did you mean to use the `<=` operator? In the latter case, ceiling will make a difference.

jaj22 replied:
Partly historical, as an earlier version lacked the floor on `_totalItems`. I think it still clarifies the intent of integer vs integer comparison, although I don't really care either way.

CalebSerafin commented on A3-Antistasi/functions/Base/fn_economicsAI.sqf:
```diff
@@ -1,67 +1,84 @@
-        };
+    if (_totalItems < ceil (_typeCoeff * _baseCap)) then {
+        private _type = selectRandom _types;
+        private _currentItems = timer getVariable [_type, 0];
+        timer setVariable [_type, _currentItems + _typeCoeff * _increase, true];
```
Note that the cap is not applied correctly. It only caps the input to this function, not the output. So the output could exceed the cap by **almost** `_typeCoeff * _increase`. However, they will randomly distribute their real cap randomly between `>= cap` and `< cap + increase`.

Even though it's incredibly unlikely and will usually have a small effect, this is valuable to parallel editing of the type variable. We can implement scheduler safety to the incrementation of the type variable by wrapping the get and set lines with `isNil {}`.

jaj22 replied:
The first point is semi-intentional. It seemed a little weird to block the overshoot here given that with multiple types, there's already a large fractional overshoot, which is actually necessary so that you don't get a giant gap after wrecking the vehicle supply. It's arguable, but it didn't seem worth the additional code.

The bigger problem with the timer variables is that other reads and even writes don't necessarily run on the server. So you could be very careful about the server-side scheduler safety and then the vars get blatted from a distance. There are improvements that could be made here (removeVehFromPool should almost certainly be a server function), but I'm not sure it's worth broadening the PR when the effects are tiny and the code is temporary.

CalebSerafin reviewed (changes requested):
Okay, pretty good changes as they make the function more readable. There are a few comments to improve the code.

jaj22 commented:
Something I should have noted: This is strictly a short-term PR. I think economicsAI was actually a bad idea and prone to counter-intuitive balance distortions even when it's working "correctly". I'm intending to strip it out in favour of a resource point system fairly soon, but that's a far more complex operation and we kinda needed a short term rebalance/bugfix. Anyway...

jaj22 commented on A3-Antistasi/functions/Base/fn_economicsAI.sqf:
```diff
@@ -1,67 +1,84 @@
+// Coeff 1.0 means one vehicle per hour with 9 players @ tierWar 7, or two vehicles per hour for 26 players.
+
+// 9 players @ tierWar 7 => balanceScale 1
+private _playerScale = (8 + count (allPlayers - entities "HeadlessClient_F")) / 17;
```
It's complex. The constant has the "real" meaning that the first player is also the commander of up to 10 HC squads and innumerable garrison statics, and so is worth a lot more, at least from the midgame onwards. Additionally, servers with fewer players tend to run lower difficulty. However, rebel AIs are so poor at dealing with armoured ground vehicles that it's hard to see a single player defeating half as many vehicles as nine players. Still, it's a much larger player scaling there was previously, and I don't think we want to make it *easy* to solo.

As for the tier scaling, the vehicle selection routines attempt to use far more fancy vehicles at higher war tiers, and the provision should scale roughly proportionally so that these routines have the intended effect. Previously, enemy factions often built up a giant buffer if players slow-played the early game, but this doesn't seem like something we should be punishing.

jaj22 commented:
Things changed:
- Added input parameters for tick time and occ/inv rates. Probably shouldn't be used, but makes it more self-documenting and might be handy for testing.
- Removed unnecessary ceil on cap check.
- Doubled the gunboat caps because they were a bit miserable.
- Serialized the vehicle count updates, both for adding and removing.

CalebSerafin reviewed (approved):
Cool, I'm happy with this short-term PR.
